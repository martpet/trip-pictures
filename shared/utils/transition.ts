interface Transition {
  from: number;
  to: number;
  handler(current: number): void;
  duration?: number;
  delay?: number;
}

export function transition({
  from,
  to,
  handler,
  duration = 150,
  delay = 20,
}: Transition) {
  return new Promise<void>((resolve) => {
    const delta = to - from;
    if (!delta) {
      resolve();
      return;
    }
    let current = from;
    let ticks = duration / delay;
    const step = delta / ticks;
    const timer = setInterval(() => {
      current += step;
      handler(current);
      ticks--;
      if (ticks <= 0) {
        clearInterval(timer);
        resolve();
      }
    }, delay);
  });
}
