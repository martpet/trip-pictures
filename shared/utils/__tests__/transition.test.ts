import { transition } from '../transition';

const handler = vi.fn();

describe('transition', () => {
  it('calls handler n times', async () => {
    await transition({
      from: 1,
      to: 4,
      duration: 100,
      delay: 10,
      handler,
    });
    expect(vi.mocked(handler).mock.calls).toMatchSnapshot();
  });

  it('resolves without calling handler if delta is 0', async () => {
    await transition({ from: 3, to: 3, handler });
    expect(vi.mocked(handler)).not.toBeCalled();
  });
});
