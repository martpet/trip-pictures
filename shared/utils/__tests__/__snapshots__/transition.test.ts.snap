// Vitest Snapshot v1

exports[`transition > calls handler n times 1`] = `
[
  [
    1.3,
  ],
  [
    1.6,
  ],
  [
    1.9000000000000001,
  ],
  [
    2.2,
  ],
  [
    2.5,
  ],
  [
    2.8,
  ],
  [
    3.0999999999999996,
  ],
  [
    3.3999999999999995,
  ],
  [
    3.6999999999999993,
  ],
  [
    3.999999999999999,
  ],
]
`;
